From 3c2c99474b0687ab457b5e8e439105a8963af597 Mon Sep 17 00:00:00 2001
From: Salvo Giangreco <giangrecosalvo9@gmail.com>
Date: Sat, 18 Oct 2025 17:06:50 +0200
Subject: [PATCH] Introduce HideAppListUtils

The ActivityStarter.smali portion of the upstream patch was dropped for
the One UI 9 source: its injection anchors on a mid-method control-flow
point whose compiler-assigned registers and labels drift every firmware
rebuild, and it patches services.jar (system_server), where a wrong
register bootloops the device. App hiding still works through the
ComputerEngine hooks below (shouldFilterApplication / getApplicationInfo
/ getPackageInfo), which are the primary PackageManager visibility
filter; the ActivityStarter hook was only a secondary layer. Restore it
by re-anchoring against the current ActivityStarter.smali if needed.

---
 .../ComputerEngine$$ExternalSyntheticLambda0.smali |  66 +++
 .../com/android/server/pm/ComputerEngine.smali     | 441 +++++++++++++++++++++
 2 files changed, 507 insertions(+)

diff --git a/smali_classes2/com/android/server/pm/ComputerEngine$$ExternalSyntheticLambda0.smali b/smali_classes2/com/android/server/pm/ComputerEngine$$ExternalSyntheticLambda0.smali
new file mode 100644
index 0000000..e87876c
--- /dev/null
+++ b/smali_classes2/com/android/server/pm/ComputerEngine$$ExternalSyntheticLambda0.smali
@@ -0,0 +1,66 @@
+.class public final synthetic Lcom/android/server/pm/ComputerEngine$$ExternalSyntheticLambda0;
+.super Ljava/lang/Object;
+.source "qb/106584599 c4c4375f2809767d35c3924eb0560cb9c53861c2af8e84225184923cb2856e18"
+
+# interfaces
+.implements Ljava/util/function/Predicate;
+
+
+# instance fields
+.field public final synthetic $r8$classId:I
+
+.field public final synthetic f$0:Ljava/util/Set;
+
+
+# direct methods
+.method public synthetic constructor <init>(Ljava/util/Set;I)V
+    .locals 0
+
+    invoke-direct {p0}, Ljava/lang/Object;-><init>()V
+
+    iput p2, p0, Lcom/android/server/pm/ComputerEngine$$ExternalSyntheticLambda0;->$r8$classId:I
+
+    iput-object p1, p0, Lcom/android/server/pm/ComputerEngine$$ExternalSyntheticLambda0;->f$0:Ljava/util/Set;
+
+    return-void
+.end method
+
+
+# virtual methods
+.method public final test(Ljava/lang/Object;)Z
+    .locals 1
+
+    iget v0, p0, Lcom/android/server/pm/ComputerEngine$$ExternalSyntheticLambda0;->$r8$classId:I
+
+    packed-switch v0, :pswitch_data_0
+
+    iget-object p0, p0, Lcom/android/server/pm/ComputerEngine$$ExternalSyntheticLambda0;->f$0:Ljava/util/Set;
+
+    check-cast p1, Landroid/content/pm/PackageInfo;
+
+    iget-object p1, p1, Landroid/content/pm/PackageInfo;->packageName:Ljava/lang/String;
+
+    invoke-interface {p0, p1}, Ljava/util/Set;->contains(Ljava/lang/Object;)Z
+
+    move-result p0
+
+    return p0
+
+    :pswitch_0
+    iget-object p0, p0, Lcom/android/server/pm/ComputerEngine$$ExternalSyntheticLambda0;->f$0:Ljava/util/Set;
+
+    check-cast p1, Landroid/content/pm/ApplicationInfo;
+
+    iget-object p1, p1, Landroid/content/pm/ApplicationInfo;->packageName:Ljava/lang/String;
+
+    invoke-interface {p0, p1}, Ljava/util/Set;->contains(Ljava/lang/Object;)Z
+
+    move-result p0
+
+    return p0
+
+    :pswitch_data_0
+    .packed-switch 0x0
+        :pswitch_0
+    .end packed-switch
+.end method
diff --git a/smali_classes2/com/android/server/pm/ComputerEngine.smali b/smali_classes2/com/android/server/pm/ComputerEngine.smali
index acbb4c9..f357d0c 100644
--- a/smali_classes2/com/android/server/pm/ComputerEngine.smali
+++ b/smali_classes2/com/android/server/pm/ComputerEngine.smali
@@ -583,6 +583,20 @@
     return-void
 .end method
 
+.method public static isBootCompleted()Z
+    .locals 2
+
+    const-string/jumbo v0, "sys.boot_completed"
+
+    const/4 v1, 0x0
+
+    invoke-static {v0, v1}, Landroid/os/SystemProperties;->getBoolean(Ljava/lang/String;Z)Z
+
+    move-result v0
+
+    return v0
+.end method
+
 .method public static resolveExternalPackageName(Lcom/android/server/pm/pkg/AndroidPackage;)Ljava/lang/String;
     .locals 1
 
@@ -1545,6 +1559,144 @@
     return p0
 .end method
 
+.method public final canHideApp(ILjava/lang/String;)Z
+    .locals 4
+
+    invoke-static {}, Lcom/android/server/pm/ComputerEngine;->isBootCompleted()Z
+
+    move-result v0
+
+    const/4 v1, 0x0
+
+    if-eqz v0, :cond_6
+
+    iget-object v0, p0, Lcom/android/server/pm/ComputerEngine;->mContext:Landroid/content/Context;
+
+    if-eqz v0, :cond_6
+
+    invoke-virtual {v0}, Landroid/content/Context;->getPackageManager()Landroid/content/pm/PackageManager;
+
+    move-result-object v0
+
+    if-nez v0, :cond_0
+
+    goto :goto_0
+
+    :cond_0
+    iget-object v0, p0, Lcom/android/server/pm/ComputerEngine;->mContext:Landroid/content/Context;
+
+    invoke-virtual {v0}, Landroid/content/Context;->getPackageManager()Landroid/content/pm/PackageManager;
+
+    move-result-object v0
+
+    invoke-virtual {v0, p1}, Landroid/content/pm/PackageManager;->getNameForUid(I)Ljava/lang/String;
+
+    move-result-object v0
+
+    if-eqz v0, :cond_6
+
+    invoke-static {v0}, Landroid/text/TextUtils;->isEmpty(Ljava/lang/CharSequence;)Z
+
+    move-result v2
+
+    if-eqz v2, :cond_1
+
+    goto :goto_0
+
+    :cond_1
+    const-string v2, "com.android.vending"
+
+    invoke-virtual {v0, v2}, Ljava/lang/String;->contains(Ljava/lang/CharSequence;)Z
+
+    move-result v2
+
+    const/4 v3, 0x1
+
+    if-eqz v2, :cond_2
+
+    return v3
+
+    :cond_2
+    if-eqz p2, :cond_6
+
+    invoke-static {p2}, Landroid/text/TextUtils;->isEmpty(Ljava/lang/CharSequence;)Z
+
+    move-result v2
+
+    if-eqz v2, :cond_3
+
+    goto :goto_0
+
+    :cond_3
+    invoke-virtual {v0, p2}, Ljava/lang/String;->contains(Ljava/lang/CharSequence;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_4
+
+    return v1
+
+    :cond_4
+    const-string/jumbo v0, "youtube"
+
+    invoke-virtual {p2, v0}, Ljava/lang/String;->contains(Ljava/lang/CharSequence;)Z
+
+    move-result v0
+
+    if-nez v0, :cond_6
+
+    const-string/jumbo v0, "microg"
+
+    invoke-virtual {p2, v0}, Ljava/lang/String;->contains(Ljava/lang/CharSequence;)Z
+
+    move-result v0
+
+    if-nez v0, :cond_6
+
+    const-string/jumbo v0, "revanced"
+
+    invoke-virtual {p2, v0}, Ljava/lang/String;->contains(Ljava/lang/CharSequence;)Z
+
+    move-result v0
+
+    if-nez v0, :cond_6
+
+    const-string v0, "gms"
+
+    invoke-virtual {p2, v0}, Ljava/lang/String;->contains(Ljava/lang/CharSequence;)Z
+
+    move-result p2
+
+    if-eqz p2, :cond_5
+
+    goto :goto_0
+
+    :cond_5
+    invoke-virtual {p0, p1}, Lcom/android/server/pm/ComputerEngine;->isCallerSystem(I)Z
+
+    move-result p0
+
+    if-nez p0, :cond_6
+
+    invoke-static {p1}, Landroid/os/Process;->isIsolated(I)Z
+
+    move-result p0
+
+    if-nez p0, :cond_6
+
+    invoke-static {p1}, Landroid/os/Process;->isSdkSandboxUid(I)Z
+
+    move-result p0
+
+    if-nez p0, :cond_6
+
+    return v3
+
+    :cond_6
+    :goto_0
+    return v1
+.end method
+
 .method public final canPackageQuery(Ljava/lang/String;[Ljava/lang/String;I)[Z
     .locals 10
 
@@ -12380,6 +12532,29 @@
 
     invoke-static {}, Landroid/os/Binder;->getCallingUid()I
 
+    move-result v0
+
+    invoke-virtual {p0, v0, p1}, Lcom/android/server/pm/ComputerEngine;->canHideApp(ILjava/lang/String;)Z
+
+    move-result v0
+
+    if-eqz v0, :hma_get_app_visible
+
+    iget-object v0, p0, Lcom/android/server/pm/ComputerEngine;->mContext:Landroid/content/Context;
+
+    invoke-static {v0, p1}, Lio/mesalabs/unica/HideAppListUtils;->shouldHideAppList(Landroid/content/Context;Ljava/lang/String;)Z
+
+    move-result v0
+
+    if-eqz v0, :hma_get_app_visible
+
+    const/4 v0, 0x0
+
+    return-object v0
+
+    :hma_get_app_visible
+    invoke-static {}, Landroid/os/Binder;->getCallingUid()I
+
     move-result v1
 
     move-object v0, p0
@@ -16649,6 +16824,29 @@
 .method public final getPackageInfo(Ljava/lang/String;JI)Landroid/content/pm/PackageInfo;
     .locals 8
 
+    invoke-static {}, Landroid/os/Binder;->getCallingUid()I
+
+    move-result v0
+
+    invoke-virtual {p0, v0, p1}, Lcom/android/server/pm/ComputerEngine;->canHideApp(ILjava/lang/String;)Z
+
+    move-result v0
+
+    if-eqz v0, :hma_get_pkg_visible
+
+    iget-object v0, p0, Lcom/android/server/pm/ComputerEngine;->mContext:Landroid/content/Context;
+
+    invoke-static {v0, p1}, Lio/mesalabs/unica/HideAppListUtils;->shouldHideAppList(Landroid/content/Context;Ljava/lang/String;)Z
+
+    move-result v0
+
+    if-eqz v0, :hma_get_pkg_visible
+
+    const/4 v0, 0x0
+
+    return-object v0
+
+    :hma_get_pkg_visible
     const-wide/16 v4, -0x1
 
     invoke-static {}, Landroid/os/Binder;->getCallingUid()I
@@ -21103,6 +21301,88 @@
     return p0
 .end method
 
+.method public final isCallerHome(II)Z
+    .locals 2
+
+    iget-object v0, p0, Lcom/android/server/pm/ComputerEngine;->mDefaultAppProvider:Lcom/android/server/pm/DefaultAppProvider;
+
+    const-string v1, "android.app.role.HOME"
+
+    invoke-virtual {v0, p2, v1}, Lcom/android/server/pm/DefaultAppProvider;->getRoleHolder(ILjava/lang/String;)Ljava/lang/String;
+
+    move-result-object p2
+
+    if-nez p2, :cond_0
+
+    const/4 p0, 0x0
+
+    return p0
+
+    :cond_0
+    invoke-virtual {p0, p1, p2}, Lcom/android/server/pm/ComputerEngine;->isCallerSameApp(ILjava/lang/String;)Z
+
+    move-result p0
+
+    return p0
+.end method
+
+.method public final isCallerSystem(I)Z
+    .locals 2
+
+    invoke-static {p1}, Lcom/android/server/pm/PackageManagerServiceUtils;->isSystemOrRootOrShell(I)Z
+
+    move-result v0
+
+    const/4 v1, 0x1
+
+    if-eqz v0, :cond_0
+
+    return v1
+
+    :cond_0
+    iget-object p0, p0, Lcom/android/server/pm/ComputerEngine;->mSettings:Lcom/android/server/pm/ComputerEngine$Settings;
+
+    invoke-static {p1}, Landroid/os/UserHandle;->getAppId(I)I
+
+    move-result p1
+
+    iget-object p0, p0, Lcom/android/server/pm/ComputerEngine$Settings;->mSettings:Lcom/android/server/pm/Settings;
+
+    invoke-virtual {p0, p1}, Lcom/android/server/pm/Settings;->getSettingLPr(I)Lcom/android/server/pm/SettingBase;
+
+    move-result-object p0
+
+    const/4 p1, 0x0
+
+    if-nez p0, :cond_1
+
+    return p1
+
+    :cond_1
+    invoke-virtual {p0}, Lcom/android/server/pm/SettingBase;->getFlags()I
+
+    move-result p0
+
+    and-int/lit8 v0, p0, 0x1
+
+    if-eq v0, v1, :cond_3
+
+    const/16 v0, 0x80
+
+    and-int/2addr p0, v0
+
+    if-ne p0, v0, :cond_2
+
+    goto :goto_0
+
+    :cond_2
+    return p1
+
+    :cond_3
+    :goto_0
+    return v1
+.end method
+
 .method public final isComponentVisibleToInstantApp(ILandroid/content/ComponentName;)Z
     .locals 6
 
@@ -27627,6 +27907,82 @@
     return-void
 .end method
 
+.method public recreateApplicationList(ILandroid/content/Context;ILjava/util/List;)Ljava/util/List;
+    .locals 0
+
+    new-instance p3, Ljava/util/ArrayList;
+
+    invoke-direct {p3, p4}, Ljava/util/ArrayList;-><init>(Ljava/util/Collection;)V
+
+    const/4 p4, 0x0
+
+    invoke-virtual {p0, p1, p4}, Lcom/android/server/pm/ComputerEngine;->canHideApp(ILjava/lang/String;)Z
+
+    move-result p0
+
+    if-nez p0, :cond_0
+
+    return-object p3
+
+    :cond_0
+    invoke-static {p2}, Lio/mesalabs/unica/HideAppListUtils;->getApps(Landroid/content/Context;)Ljava/util/Set;
+
+    move-result-object p0
+
+    new-instance p1, Lcom/android/server/pm/ComputerEngine$$ExternalSyntheticLambda0;
+
+    invoke-direct {p1, p0, p4}, Lcom/android/server/pm/ComputerEngine$$ExternalSyntheticLambda0;-><init>(Ljava/util/Set;I)V
+
+    invoke-interface {p3, p1}, Ljava/util/List;->removeIf(Ljava/util/function/Predicate;)Z
+
+    return-object p3
+.end method
+
+.method public recreatePackageList(ILandroid/content/Context;ILandroid/content/pm/ParceledListSlice;)Landroid/content/pm/ParceledListSlice;
+    .locals 0
+
+    new-instance p3, Ljava/util/ArrayList;
+
+    invoke-virtual {p4}, Landroid/content/pm/ParceledListSlice;->getList()Ljava/util/List;
+
+    move-result-object p4
+
+    invoke-direct {p3, p4}, Ljava/util/ArrayList;-><init>(Ljava/util/Collection;)V
+
+    const/4 p4, 0x0
+
+    invoke-virtual {p0, p1, p4}, Lcom/android/server/pm/ComputerEngine;->canHideApp(ILjava/lang/String;)Z
+
+    move-result p0
+
+    if-nez p0, :cond_0
+
+    new-instance p0, Landroid/content/pm/ParceledListSlice;
+
+    invoke-direct {p0, p3}, Landroid/content/pm/ParceledListSlice;-><init>(Ljava/util/List;)V
+
+    return-object p0
+
+    :cond_0
+    invoke-static {p2}, Lio/mesalabs/unica/HideAppListUtils;->getApps(Landroid/content/Context;)Ljava/util/Set;
+
+    move-result-object p0
+
+    new-instance p1, Lcom/android/server/pm/ComputerEngine$$ExternalSyntheticLambda0;
+
+    const/4 p2, 0x1
+
+    invoke-direct {p1, p0, p2}, Lcom/android/server/pm/ComputerEngine$$ExternalSyntheticLambda0;-><init>(Ljava/util/Set;I)V
+
+    invoke-interface {p3, p1}, Ljava/util/List;->removeIf(Ljava/util/function/Predicate;)Z
+
+    new-instance p0, Landroid/content/pm/ParceledListSlice;
+
+    invoke-direct {p0, p3}, Landroid/content/pm/ParceledListSlice;-><init>(Ljava/util/List;)V
+
+    return-object p0
+.end method
+
 .method public resolveComponentName()Landroid/content/ComponentName;
     .locals 0
 
@@ -28289,6 +28645,17 @@
 .method public final shouldFilterApplication(Lcom/android/server/pm/pkg/PackageStateInternal;ILandroid/content/ComponentName;IIZZ)Z
     .locals 8
 
+    invoke-virtual {p0, p1, p2, p5}, Lcom/android/server/pm/ComputerEngine;->shouldFilterApplicationCustom(Lcom/android/server/pm/pkg/PackageStateInternal;II)Z
+
+    move-result v0
+
+    if-eqz v0, :hma_filter_continue
+
+    const/4 v0, 0x1
+
+    return v0
+
+    :hma_filter_continue
     invoke-static {p2}, Landroid/os/Process;->isSdkSandboxUid(I)Z
 
     move-result v0
@@ -28567,6 +28934,82 @@
     return v0
 .end method
 
+.method public final shouldFilterApplicationCustom(Lcom/android/server/pm/pkg/PackageStateInternal;II)Z
+    .locals 2
+
+    invoke-static {}, Lcom/android/server/pm/ComputerEngine;->isBootCompleted()Z
+
+    move-result v0
+
+    const/4 v1, 0x0
+
+    if-nez v0, :cond_0
+
+    return v1
+
+    :cond_0
+    if-nez p1, :cond_1
+
+    return v1
+
+    :cond_1
+    invoke-interface {p1}, Lcom/android/server/pm/pkg/PackageState;->getPackageName()Ljava/lang/String;
+
+    move-result-object p1
+
+    if-nez p1, :cond_2
+
+    return v1
+
+    :cond_2
+    invoke-virtual {p0, p2, p1}, Lcom/android/server/pm/ComputerEngine;->isCallerSameApp(ILjava/lang/String;)Z
+
+    move-result v0
+
+    if-nez v0, :cond_4
+
+    invoke-virtual {p0, p2}, Lcom/android/server/pm/ComputerEngine;->isCallerSystem(I)Z
+
+    move-result v0
+
+    if-nez v0, :cond_4
+
+    invoke-virtual {p0, p2, p3}, Lcom/android/server/pm/ComputerEngine;->isCallerHome(II)Z
+
+    move-result p2
+
+    if-eqz p2, :cond_3
+
+    goto :goto_0
+
+    :cond_3
+    invoke-static {}, Landroid/os/Binder;->getCallingUid()I
+
+    move-result p2
+
+    invoke-virtual {p0, p2, p1}, Lcom/android/server/pm/ComputerEngine;->canHideApp(ILjava/lang/String;)Z
+
+    move-result p2
+
+    if-eqz p2, :cond_4
+
+    iget-object p0, p0, Lcom/android/server/pm/ComputerEngine;->mContext:Landroid/content/Context;
+
+    invoke-static {p0, p1}, Lio/mesalabs/unica/HideAppListUtils;->shouldHideAppList(Landroid/content/Context;Ljava/lang/String;)Z
+
+    move-result p0
+
+    if-eqz p0, :cond_4
+
+    const/4 p0, 0x1
+
+    return p0
+
+    :cond_4
+    :goto_0
+    return v1
+.end method
+
 .method public final shouldFilterApplicationIncludingUninstalled(Lcom/android/server/pm/SharedUserSetting;II)Z
     .locals 2
 
-- 
2.51.2
